src/python/pants/source/wrapped_globs.py

```diff
@@ -120,6 +125,13 @@ def files_hash(self):
         h.update(f.read())
     return h.digest()
 
+  def iter_relative_paths(self):
+    """An alternative `__iter__` that joins files with the relative root."""
+    for f in self:
+      path_from_buildroot = os.path.join(self.rel_root, f)
+      if matches_filespec(path_from_buildroot, self.filespec):
```

Comment:
This is redundant. The files in the collection are already guaranteed to match the filespec: the filespec is just metadata about how the fileset was constructed.

The point of the `SpecAddressMapper` is to _not_ expand the fileset, and to just use the `filespec` instead. This was based on the assumption that all `FilesetWithSpec` implementations were lazy.

What I was suggesting was an abstract method on `FilesetWithSpec` something like:
```
def matches(self, paths):
    """Returns the from the given set that match this FilesetWithSpec."""
```
...and the implementation for `Eager` would be to just look in the fileset, while the implementation for `Lazy` would be to compile the regex from the filespec and use that.